sequence_task_1/student/include/impl.h

```diff
@@ -1,51 +1,86 @@
 #pragma once
+#include <stdexcept>
 #include <string>
 
-struct Song
-{
-    std::string name;
-    explicit Song(std::string song_name)
-        : name {std::move(song_name)}
-    {
-    }
+struct Song {
+  std::string name;
+  explicit Song(std::string song_name) : name{std::move(song_name)} {}
 };
 
-bool operator==(const Song& lhs, const Song& rhs)
-{
-    return lhs.name == rhs.name;
+bool operator==(const Song& lhs, const Song& rhs) {
+  return lhs.name == rhs.name;
 }
 
-template<class Container>
-class StaticPlaylist
-{
-public:
-    /** @todo Member traits */
+template <class Container>
+class StaticPlaylist {
+ public:
+  /** @todo Member traits */
+  using size_type = typename Container::size_type;
+  using value_type = typename Container::value_type;
+  using reference = typename Container::reference;
+  using const_reference = typename Container::const_reference;
+  using iterator = typename Container::iterator;
+  using const_iterator = typename Container::const_iterator;
+  using difference_type = typename Container::difference_type;
+
+  /** @todo Iterators */
+  const_iterator begin() {
+    return m_tracklist.begin();
+  }
 
-    /** @todo Iterators */
+  const_iterator end() {
+    return m_tracklist.end();
+  }
 
-    /** @todo Constructor from any reversible sequence container */
+  /** @todo Constructor from any reversible sequence container */
+  template <class T>
+  StaticPlaylist(const T& track_list) {
+    m_tracklist.assign(track_list.begin(), track_list.end());
+  }
 
-    /** @todo Assignment from any reversible sequence container */
+  /** @todo Assignment from any reversible sequence container */
+  template <class T>
+  StaticPlaylist& operator=(const T& track_list) {
+    m_tracklist.assign(track_list.begin(), track_list.end());
+    return *this;
+  }
 
-    /** @todo Add track from initializer */
-    template<class... Args>
-    const Song& play(Args&&... songData);
+  /** @todo Add track from initializer */
+  template <class... Args>
+  const Song& play(Args&&... songData) {
+    m_tracklist.emplace_back(songData...);
+    return m_tracklist.back();
+  }
 
-    /** @todo Add track */
-    const Song& play(const Song& song);
+  /** @todo Add track */
+  const Song& play(const Song& song) {
+    m_tracklist.push_back(song);
+    return m_tracklist.back();
+  }
 
-    /** @todo Get first track in playlist stack */
-    const Song& current() const;
+  /** @todo Get first track in playlist stack */
+  const Song& current() const {
+    return m_tracklist.front();
+  }
 
-    /** @todo Skip to the next track in playlist, remove current */
-    void switchNext();
+  /** @todo Skip to the next track in playlist, remove current */
+  void switchNext() {
+    if (m_tracklist.empty()) {
+      throw std::out_of_range("");
+    }
+    m_tracklist.erase(m_tracklist.begin());
```

Comment:
**Not effective for vector!**

Reply:
Can't find a way to do this more efficiently without losing the songs order.

Reply:
What about the reverse iterators? :)
Maybe we need a special way to copy the input data.